File: src/support/colormap.c
#include "sounds/colormap.h"

#include <stddef.h>

typedef struct SoundColormapData {
    const char *name;
    const SoundColor *stops;
    int stop_count;
} SoundColormapData;

static const SoundColor viridis_stops[] = {
    {0.267004F, 0.004874F, 0.329415F},
    {0.282623F, 0.140926F, 0.457517F},
    {0.253935F, 0.265254F, 0.529983F},
    {0.206756F, 0.371758F, 0.553117F},
    {0.163625F, 0.471133F, 0.558148F},
    {0.127568F, 0.566949F, 0.550556F},
    {0.134692F, 0.658636F, 0.517649F},
    {0.266941F, 0.748751F, 0.440573F},
    {0.477504F, 0.821444F, 0.318195F},
    {0.741388F, 0.873449F, 0.149561F},
    {0.993248F, 0.906157F, 0.143936F},
};

static const SoundColor magma_stops[] = {
    {0.001462F, 0.000466F, 0.013866F},
    {0.078815F, 0.054184F, 0.211667F},
    {0.232077F, 0.059889F, 0.437695F},
    {0.390384F, 0.100379F, 0.501864F},
    {0.550287F, 0.161158F, 0.505719F},
    {0.716387F, 0.214982F, 0.475290F},
    {0.868793F, 0.287728F, 0.409303F},
    {0.967671F, 0.439703F, 0.359810F},
    {0.994738F, 0.624350F, 0.427397F},
    {0.995680F, 0.812706F, 0.572645F},
    {0.987053F, 0.991438F, 0.749504F},
};

static const SoundColor inferno_stops[] = {
    {0.001462F, 0.000466F, 0.013866F},
    {0.087411F, 0.044556F, 0.224813F},
    {0.258234F, 0.038571F, 0.406485F},
    {0.416331F, 0.090203F, 0.432943F},
    {0.578304F, 0.148039F, 0.404411F},
    {0.735683F, 0.215906F, 0.330245F},
    {0.865006F, 0.316822F, 0.226055F},
    {0.954506F, 0.468744F, 0.099874F},
    {0.987622F, 0.645320F, 0.039886F},
    {0.964394F, 0.843848F, 0.273391F},
    {0.988362F, 0.998364F, 0.644924F},
};

static const SoundColor plasma_stops[] = {
    {0.050383F, 0.029803F, 0.527975F},
    {0.254627F, 0.013882F, 0.615419F},
    {0.417642F, 0.000564F, 0.658390F},
    {0.562738F, 0.051545F, 0.641509F},
    {0.692840F, 0.165141F, 0.564522F},
    {0.798216F, 0.280197F, 0.469538F},
    {0.881443F, 0.392529F, 0.383229F},
    {0.949217F, 0.517763F, 0.295662F},
    {0.988260F, 0.652325F, 0.211364F},
    {0.988648F, 0.809579F, 0.145357F},
    {0.940015F, 0.975158F, 0.131326F},
};

static const SoundColor cividis_stops[] = {
    {0.000000F, 0.135112F, 0.304751F},
    {0.032110F, 0.201199F, 0.440785F},
    {0.208926F, 0.272546F, 0.424809F},
    {0.309601F, 0.340399F, 0.424790F},
    {0.401418F, 0.411790F, 0.440708F},
    {0.488697F, 0.485318F, 0.471008F},
    {0.582087F, 0.558670F, 0.468118F},
    {0.683950F, 0.638793F, 0.444251F},
    {0.785965F, 0.720438F, 0.399613F},
    {0.896818F, 0.811030F, 0.320982F},
    {0.995737F, 0.909344F, 0.217772F},
};

static const SoundColor turbo_stops[] = {
    {0.189950F, 0.071760F, 0.232170F},
    {0.269670F, 0.348780F, 0.796310F},
    {0.244270F, 0.609370F, 0.996970F},
    {0.097500F, 0.837140F, 0.803420F},
    {0.275970F, 0.970920F, 0.516530F},
    {0.643620F, 0.989990F, 0.233560F},
    {0.883310F, 0.865530F, 0.217190F},
    {0.996070F, 0.642770F, 0.191650F},
    {0.940840F, 0.355660F, 0.070310F},
    {0.764760F, 0.143740F, 0.010410F},
    {0.479600F, 0.015830F, 0.010550F},
};

#define STOP_COUNT(stops) ((int)(sizeof(stops) / sizeof((stops)[0])))

static const SoundColormapData colormaps[SOUND_COLORMAP_COUNT] = {
    [SOUND_COLORMAP_VIRIDIS] = {"viridis", viridis_stops, STOP_COUNT(viridis_stops)},
    [SOUND_COLORMAP_MAGMA] = {"magma", magma_stops, STOP_COUNT(magma_stops)},
    [SOUND_COLORMAP_INFERNO] = {"inferno", inferno_stops, STOP_COUNT(inferno_stops)},
    [SOUND_COLORMAP_PLASMA] = {"plasma", plasma_stops, STOP_COUNT(plasma_stops)},
    [SOUND_COLORMAP_CIVIDIS] = {"cividis", cividis_stops, STOP_COUNT(cividis_stops)},
    [SOUND_COLORMAP_TURBO] = {"turbo", turbo_stops, STOP_COUNT(turbo_stops)},
};

static float clamp_unit(float value) {
    if (value < 0.0F) {
        return 0.0F;
    }

    if (value > 1.0F) {
        return 1.0F;
    }

    return value;
}

static float mix_channel(float a, float b, float amount) {
    return a + (b - a) * amount;
}

static const SoundColormapData *colormap_data(SoundColormap colormap) {
    if (colormap < 0 || colormap >= SOUND_COLORMAP_COUNT) {
        return &colormaps[SOUND_COLORMAP_VIRIDIS];
    }

    return &colormaps[colormap];
}

static SoundColor sample_stops(
    const SoundColor *stops,
    int stop_count,
    float unit
) {
    float t = clamp_unit(unit);
    float scaled = t * (float)(stop_count - 1);
    int index = (int)scaled;

    if (index >= stop_count - 1) {
        return stops[stop_count - 1];
    }

    SoundColor low = stops[index];
    SoundColor high = stops[index + 1];
    float amount = scaled - (float)index;

    return (SoundColor){
        .red = mix_channel(low.red, high.red, amount),
        .green = mix_channel(low.green, high.green, amount),
        .blue = mix_channel(low.blue, high.blue, amount),
    };
}

const char *sound_colormap_name(SoundColormap colormap) {
    return colormap_data(colormap)->name;
}

SoundColor sound_colormap_sample(SoundColormap colormap, float unit) {
    const SoundColormapData *data = colormap_data(colormap);
    return sample_stops(data->stops, data->stop_count, unit);
}

SoundColor sound_colormap_viridis(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_VIRIDIS, unit);
}

SoundColor sound_colormap_magma(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_MAGMA, unit);
}

SoundColor sound_colormap_inferno(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_INFERNO, unit);
}

SoundColor sound_colormap_plasma(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_PLASMA, unit);
}

SoundColor sound_colormap_cividis(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_CIVIDIS, unit);
}

SoundColor sound_colormap_turbo(float unit) {
    return sound_colormap_sample(SOUND_COLORMAP_TURBO, unit);
}

File: src/support/colormap.zig
//! Perceptual color maps for spectrogram rendering.

const std = @import("std");

pub const Color = struct {
    red: f32,
    green: f32,
    blue: f32,
};

pub const Colormap = enum {
    viridis,
    magma,
    inferno,
    plasma,
    cividis,
    turbo,

    pub fn name(self: Colormap) []const u8 {
        return names[@intFromEnum(self)];
    }

    pub fn index(self: Colormap) usize {
        return @intFromEnum(self);
    }

    pub fn sample(self: Colormap, unit: f32) Color {
        return sampleStops(&stop_tables[@intFromEnum(self)], unit);
    }
};

pub const order = [_]Colormap{
    .viridis,
    .magma,
    .inferno,
    .plasma,
    .cividis,
    .turbo,
};

pub const count = order.len;

pub fn at(index: isize) Colormap {
    if (index < 0) return .viridis;

    const i: usize = @intCast(index);
    if (i >= order.len) return .viridis;
    return order[i];
}

pub fn fromInt(value: i32) ?Colormap {
    return switch (value) {
        0 => .viridis,
        1 => .magma,
        2 => .inferno,
        3 => .plasma,
        4 => .cividis,
        5 => .turbo,
        else => null,
    };
}

const names = [_][]const u8{
    "viridis",
    "magma",
    "inferno",
    "plasma",
    "cividis",
    "turbo",
};

const stop_count = 11;

inline fn rgb(red: f32, green: f32, blue: f32) Color {
    return .{ .red = red, .green = green, .blue = blue };
}

const viridis_stops = [stop_count]Color{
    rgb(0.267004, 0.004874, 0.329415),
    rgb(0.282623, 0.140926, 0.457517),
    rgb(0.253935, 0.265254, 0.529983),
    rgb(0.206756, 0.371758, 0.553117),
    rgb(0.163625, 0.471133, 0.558148),
    rgb(0.127568, 0.566949, 0.550556),
    rgb(0.134692, 0.658636, 0.517649),
    rgb(0.266941, 0.748751, 0.440573),
    rgb(0.477504, 0.821444, 0.318195),
    rgb(0.741388, 0.873449, 0.149561),
    rgb(0.993248, 0.906157, 0.143936),
};

const magma_stops = [stop_count]Color{
    rgb(0.001462, 0.000466, 0.013866),
    rgb(0.078815, 0.054184, 0.211667),
    rgb(0.232077, 0.059889, 0.437695),
    rgb(0.390384, 0.100379, 0.501864),
    rgb(0.550287, 0.161158, 0.505719),
    rgb(0.716387, 0.214982, 0.475290),
    rgb(0.868793, 0.287728, 0.409303),
    rgb(0.967671, 0.439703, 0.359810),
    rgb(0.994738, 0.624350, 0.427397),
    rgb(0.995680, 0.812706, 0.572645),
    rgb(0.987053, 0.991438, 0.749504),
};

const inferno_stops = [stop_count]Color{
    rgb(0.001462, 0.000466, 0.013866),
    rgb(0.087411, 0.044556, 0.224813),
    rgb(0.258234, 0.038571, 0.406485),
    rgb(0.416331, 0.090203, 0.432943),
    rgb(0.578304, 0.148039, 0.404411),
    rgb(0.735683, 0.215906, 0.330245),
    rgb(0.865006, 0.316822, 0.226055),
    rgb(0.954506, 0.468744, 0.099874),
    rgb(0.987622, 0.645320, 0.039886),
    rgb(0.964394, 0.843848, 0.273391),
    rgb(0.988362, 0.998364, 0.644924),
};

const plasma_stops = [stop_count]Color{
    rgb(0.050383, 0.029803, 0.527975),
    rgb(0.254627, 0.013882, 0.615419),
    rgb(0.417642, 0.000564, 0.658390),
    rgb(0.562738, 0.051545, 0.641509),
    rgb(0.692840, 0.165141, 0.564522),
    rgb(0.798216, 0.280197, 0.469538),
    rgb(0.881443, 0.392529, 0.383229),
    rgb(0.949217, 0.517763, 0.295662),
    rgb(0.988260, 0.652325, 0.211364),
    rgb(0.988648, 0.809579, 0.145357),
    rgb(0.940015, 0.975158, 0.131326),
};

const cividis_stops = [stop_count]Color{
    rgb(0.000000, 0.135112, 0.304751),
    rgb(0.032110, 0.201199, 0.440785),
    rgb(0.208926, 0.272546, 0.424809),
    rgb(0.309601, 0.340399, 0.424790),
    rgb(0.401418, 0.411790, 0.440708),
    rgb(0.488697, 0.485318, 0.471008),
    rgb(0.582087, 0.558670, 0.468118),
    rgb(0.683950, 0.638793, 0.444251),
    rgb(0.785965, 0.720438, 0.399613),
    rgb(0.896818, 0.811030, 0.320982),
    rgb(0.995737, 0.909344, 0.217772),
};

const turbo_stops = [stop_count]Color{
    rgb(0.189950, 0.071760, 0.232170),
    rgb(0.269670, 0.348780, 0.796310),
    rgb(0.244270, 0.609370, 0.996970),
    rgb(0.097500, 0.837140, 0.803420),
    rgb(0.275970, 0.970920, 0.516530),
    rgb(0.643620, 0.989990, 0.233560),
    rgb(0.883310, 0.865530, 0.217190),
    rgb(0.996070, 0.642770, 0.191650),
    rgb(0.940840, 0.355660, 0.070310),
    rgb(0.764760, 0.143740, 0.010410),
    rgb(0.479600, 0.015830, 0.010550),
};

const stop_tables = [_][stop_count]Color{
    viridis_stops,
    magma_stops,
    inferno_stops,
    plasma_stops,
    cividis_stops,
    turbo_stops,
};

fn clampUnit(value: f32) f32 {
    if (value < 0.0) return 0.0;
    if (value > 1.0) return 1.0;
    return value;
}

fn mixChannel(a: f32, b: f32, amount: f32) f32 {
    return a + (b - a) * amount;
}

fn sampleStops(stops: []const Color, unit: f32) Color {
    const t = clampUnit(unit);
    const scaled = t * @as(f32, @floatFromInt(stops.len - 1));
    const index: usize = @intFromFloat(scaled);

    if (index >= stops.len - 1) return stops[stops.len - 1];

    const low = stops[index];
    const high = stops[index + 1];
    const amount = scaled - @as(f32, @floatFromInt(index));

    return .{
        .red = mixChannel(low.red, high.red, amount),
        .green = mixChannel(low.green, high.green, amount),
        .blue = mixChannel(low.blue, high.blue, amount),
    };
}

fn expectColorApprox(expected: Color, actual: Color) !void {
    try std.testing.expectApproxEqAbs(expected.red, actual.red, 1.0e-6);
    try std.testing.expectApproxEqAbs(expected.green, actual.green, 1.0e-6);
    try std.testing.expectApproxEqAbs(expected.blue, actual.blue, 1.0e-6);
}

test "colormap order, names, and index fallbacks are stable" {
    try std.testing.expectEqual(@as(usize, 6), count);
    try std.testing.expectEqual(Colormap.viridis, at(-1));
    try std.testing.expectEqual(Colormap.viridis, at(6));

    for (order, 0..) |map, i| {
        try std.testing.expectEqual(i, map.index());
        try std.testing.expectEqual(map, at(@intCast(i)));
    }

    try std.testing.expectEqualStrings("viridis", Colormap.viridis.name());
    try std.testing.expectEqualStrings("magma", Colormap.magma.name());
    try std.testing.expectEqualStrings("inferno", Colormap.inferno.name());
    try std.testing.expectEqualStrings("plasma", Colormap.plasma.name());
    try std.testing.expectEqualStrings("cividis", Colormap.cividis.name());
    try std.testing.expectEqualStrings("turbo", Colormap.turbo.name());
    try std.testing.expectEqual(@as(?Colormap, null), fromInt(-1));
    try std.testing.expectEqual(@as(?Colormap, Colormap.turbo), fromInt(5));
}

test "colormap samples clamp and interpolate between color stops" {
    try expectColorApprox(rgb(0.267004, 0.004874, 0.329415), Colormap.viridis.sample(-10.0));
    try expectColorApprox(rgb(0.993248, 0.906157, 0.143936), Colormap.viridis.sample(10.0));
    try expectColorApprox(rgb(0.987053, 0.991438, 0.749504), Colormap.magma.sample(1.0));

    try expectColorApprox(
        rgb(0.490190, 0.0260545, 0.6499495),
        Colormap.plasma.sample(0.25),
    );
    try expectColorApprox(
        rgb(0.763465, 0.927760, 0.225375),
        Colormap.turbo.sample(0.55),
    );
}
